url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/SimpleThreadPool.java

```diff
@@ -0,0 +1,567 @@
+    for (int i = 1; i <= count; ++i) {
+      WorkerThread wt = new WorkerThread(this, threadGroup, getThreadNamePrefix() + "-" + i, getThreadPriority(), isMakeThreadsDaemons());
+      if (isThreadsInheritContextClassLoaderOfInitializingThread()) {
+        wt.setContextClassLoader(Thread.currentThread().getContextClassLoader());
+      }
+      workers.add(wt);
+    }
+
+    return workers;
+  }
+
+  /**
+   * <p>
+   * Terminate any worker threads in this thread group.
+   * </p>
+   * <p>
+   * Jobs currently in progress will complete.
+   * </p>
+   */
+  public void shutdown() {
+
+    shutdown(true);
+  }
+
+  /**
+   * <p>
+   * Terminate any worker threads in this thread group.
+   * </p>
+   * <p>
+   * Jobs currently in progress will complete.
+   * </p>
+   */
+  @Override
+  public void shutdown(boolean waitForJobsToComplete) {
+
+    synchronized (nextRunnableLock) {
+      isShutdown = true;
+
+      if (workers == null) {
+        return;
+      }
+
+      // signal each worker thread to shut down
+      Iterator workerThreads = workers.iterator();
+      while (workerThreads.hasNext()) {
+        WorkerThread wt = (WorkerThread) workerThreads.next();
+        wt.shutdown();
+        availWorkers.remove(wt);
+      }
+
+      // Give waiting (wait(1000)) worker threads a chance to shut down.
+      // Active worker threads will shut down after finishing their
+      // current job.
+      nextRunnableLock.notifyAll();
+
+      if (waitForJobsToComplete == true) {
+
+        // wait for hand-off in runInThread to complete...
+        while (handoffPending) {
+          try {
+            nextRunnableLock.wait(100);
+          } catch (Throwable t) {
+          }
+        }
+
+        // Wait until all worker threads are shut down
+        while (busyWorkers.size() > 0) {
+          WorkerThread wt = busyWorkers.getFirst();
+          try {
+            getLog().debug("Waiting for thread " + wt.getName() + " to shut down");
+
+            // note: with waiting infinite time the
+            // application may appear to 'hang'.
+            nextRunnableLock.wait(2000);
+          } catch (InterruptedException ex) {
+          }
+        }
+
+        getLog().debug("shutdown complete");
+      }
+    }
+  }
+
+  /**
+   * <p>
+   * Run the given <code>Runnable</code> object in the next available <code>Thread</code>. If while waiting the thread pool is asked to shut down, the
+   * Runnable is executed immediately within a new additional thread.
+   * </p>
+   * 
+   * @param runnable the <code>Runnable</code> to be added.
+   */
+  @Override
+  public boolean runInThread(Runnable runnable) {
+
+    if (runnable == null) {
+      return false;
+    }
+
+    synchronized (nextRunnableLock) {
+
+      handoffPending = true;
+
+      // Wait until a worker thread is available
+      while ((availWorkers.size() < 1) && !isShutdown) {
+        try {
+          nextRunnableLock.wait(500);
+        } catch (InterruptedException ignore) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)